### PR TITLE
[DEVX-11] feat: Asset & AssetList class

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/__tests__/resources/asset.unit.spec.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    axiosMock
+      .onPost(new RegExp('/assets$'), {
+        items: [newRoot, ...childArray],
+      })
+      .replyOnce(200, {
+        items: [newRoot, ...childArray],
+      });
+    axiosMock
+      .onPost(new RegExp('/assets'), {
```
Comment (ghost) [inline src/__tests__/resources/asset.unit.spec.ts]:
```suggestion
      .onPost(new RegExp('/assets/list$'), {
```

File: src/standardMethods.ts [hunk 2]
```diff
@@ -348,7 +366,9 @@ export function generateSearchEndpoint<RequestParams, ResponseType>(
       },
       true
     );
-    return metadataMap.addAndReturn(response.data.items, response);
+    const { items } = response.data;
+
```
Comment (ghost) [inline src/standardMethods.ts]: Remove empty line

**Comment by ghost** [comment]
Approved from my side but can add the review to Github (I am the PR creator). @polomani review?

**Comment by polomani** [comment]
Ok, np

File: src/__tests__/resources/asset.integration.spec.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    test('subtree', async () => {
+      const newRoot = {
+        ...newRootAsset,
+        externalId: 'test-root' + randomInt(),
+      };
+      const childArray = [];
+      for (let index = 0; index < 3; index++) {
+        childArray.push({
+          ...newChildAsset,
+          externalId: 'test-child' + randomInt() + index,
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]: you don't need an index here `randomInt` is enough, isn't it?

File: src/__tests__/resources/asset.integration.spec.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          parentExternalId: newRoot.externalId,
+        });
+      }
+      const grandChildArray = [];
+      for (let index = 0; index < 105; index++) {
+        grandChildArray.push({
+          ...newGrandChildAsset,
+          parentExternalId: childArray[0].externalId,
+          externalId: 'test-grandchild' + randomInt() + index,
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]: same here

File: src/__tests__/resources/asset.integration.spec.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    externalId: 'test-root' + randomInt(),
+  };
+  const createdAssets = await client.assets.create([newRoot]);
+  const content =
+    api instanceof TimeSeriesAPI
+      ? { assetId: createdAssets[0].id }
+      : { assetIds: [createdAssets[0].id] };
+  const resourceList = new Array(1003).fill(content);
+  const resources = await api.create(resourceList);
+  await sleepPromise(10000); // eventual consistency in the backend
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]: I am not sure we can add 10sec delays anymore. `yarn test` takes a lot of time already
I think we should mock it here

File: src/__tests__/resources/asset.integration.spec.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+          parentExternalId: childArray[0].externalId,
+          externalId: 'test-grandchild' + randomInt() + index,
+        });
+      }
+      const createdAssets = await client.assets.create([
+        newRoot,
+        ...childArray,
+        ...grandChildArray,
+      ]);
+      await sleepPromise(5000); // eventual consistency in the backend
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]: fails sometimes, but we can't add even more delays

File: src/__tests__/resources/asset.integration.spec.ts [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    test('timeseries from Asset', async () => {
+      await testResourceType(client, client.timeseries, newRootAsset);
+    });
+  });
+});
+
+async function testResourceType(
+  client: CogniteClient,
+  api: TimeSeriesAPI | EventsAPI,
+  newRootAsset: any
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]: can you move `newRootAsset` one scope up so you don't need to pass it every time?

File: src/__tests__/resources/asset.integration.spec.ts [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  };
+  const createdAssets = await client.assets.create([newRoot]);
+  const content =
+    api instanceof TimeSeriesAPI
+      ? { assetId: createdAssets[0].id }
+      : { assetIds: [createdAssets[0].id] };
+  const resourceList = new Array(1003).fill(content);
+  const resources = await api.create(resourceList);
+  await sleepPromise(10000); // eventual consistency in the backend
+  let fetchedResource: GetTimeSeriesMetadataDTO[] | CogniteEvent[];
```
Comment (polomani) [inline src/__tests__/resources/asset.integration.spec.ts]:
```suggestion
  let fetchedResource: (GetTimeSeriesMetadataDTO | CogniteEvent)[];
```
by doing this you can remove `// @ts-ignore` below

File: src/__tests__/resources/assetList.integration.specs.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  });
+  test('get timeseries', async () => {
+    const assets = [{ name: 'First asset' }, { name: 'Second asset' }];
+    const createdAssets = await client.assets.create(assets);
+    const timeseries = [
+      { name: 'Pressure sensor', assetId: createdAssets[0].id },
+      { name: 'Temprature sensor', assetId: createdAssets[1].id },
+    ];
+    const createdTimeseries = await client.timeseries.create(timeseries);
+    await sleepPromise(2000); // Backend issues
```
Comment (polomani) [inline src/__tests__/resources/assetList.integration.specs.ts]: never fails?

File: src/__tests__/resources/assetList.integration.specs.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+  test('get events', async () => {
+    const assets = [{ name: 'First asset' }, { name: 'Second asset' }];
+    const createdAssets = await client.assets.create(assets);
+    const events = [
+      { description: 'Test event 1', assetIds: [createdAssets[0].id] },
+      { description: 'Test event 2', assetIds: [createdAssets[1].id] },
+    ];
+    const createdEvents = await client.events.create(events);
+    await sleepPromise(2000); // Backend issues
```
Comment (polomani) [inline src/__tests__/resources/assetList.integration.specs.ts]: 2 sec is always enough?

File: src/__tests__/standardMethods.spec.ts [hunk 11]
```diff
@@ -65,7 +65,12 @@ describe('standard methods', () => {
         axiosMock,
         metadataMap,
         'path',
-        generateCreateEndpoint(axiosInstance, 'path', metadataMap)
+        generateCreateEndpoint(
+          axiosInstance,
+          'path',
+          metadataMap,
+          items => items
```
Comment (polomani) [inline src/__tests__/standardMethods.spec.ts]: `items => items` should be removed after you fix the thing I suggested below

File: src/__tests__/standardMethods.spec.ts [hunk 12]
```diff
@@ -121,11 +136,12 @@ describe('standard methods', () => {
   });
   describe('generateListEndpoint', () => {
     function createMockedListEndpoint() {
-      const listEndpoint = generateListEndpoint<any, number>(
+      const listEndpoint = generateListEndpoint<any, number, number>(
```
Comment (polomani) [inline src/__tests__/standardMethods.spec.ts]: same here, please make default transformer inside `generateListEndpoint `

File: src/resources/3d/assetMappings3DApi.ts [hunk 13]
```diff
@@ -37,11 +37,16 @@ export class AssetMappings3DAPI {
...
+      AssetMappings3DListFilter,
+      AssetMapping3D,
+      AssetMapping3D
+    >(
       this.instance,
       parameterizePath(this.project, modelId, revisionId),
       this.map,
-      false
+      false,
+      items => items
```
Comment (polomani) [inline src/resources/3d/assetMappings3DApi.ts]: please do this in such places, then you don't need to pass `items => items` and <T, T> every time

Screenshot 2019-08-01 at 15 07 54

Comment (polomani) [reply src/resources/3d/assetMappings3DApi.ts]: but without `export`
I accidentally added it
I can help you with this one a bit later

File: src/resources/assets/assetsApi.ts [hunk 14]
```diff
@@ -90,39 +87,127 @@ export class AssetsAPI {
...
+      map,
+      transformResponse
+    );
     this.delete = generateDeleteEndpointWithParams(instance, path, map);
   }
+
+  public async retrieveSubtree(id: types.CogniteInternalId, depth: number) {
+    const currentDepth: number = 0;
+    const rootAssetList = await this.retrieve([{ id }]);
+    return this.getAssetSubtree(rootAssetList, currentDepth, depth);
```
Comment (polomani) [inline src/resources/assets/assetsApi.ts:148]: maybe you don't need `currentDepth ` variable? just pass 0?

File: src/resources/assets/assetsApi.ts [hunk 15]
```diff
@@ -90,39 +87,127 @@ export class AssetsAPI {
...
 
-export type AssetUpdateEndpoint = (changes: AssetChange[]) => Promise<Asset[]>;
+export type AssetUpdateEndpoint = (
+  changes: types.AssetChange[]
+) => Promise<AssetList>;
 
 export type AssetSearchEndpoint = (
-  query: AssetSearchFilter
-) => Promise<Asset[]>;
+  query: types.AssetSearchFilter
```
Comment (polomani) [inline src/resources/assets/assetsApi.ts]: you can import all types from types.ts, but do `Asset as AssetProps`, like
```
import { ExternalAssetItem, AssetListScope, AssetIdEither, AssetChange, AssetSearchFilter, CogniteInternalId, Asset as TypeAsset } from '../../types/types';
```
and only asset class separately
```
import { Asset } from '../classes/asset';
```

and then
```
... query: AssetSearchFilter
```
```
private transformToAssetListObject = (assets: AssetProps[]) => {
```

So you don't have ugly `types.Something`  everywhere

Comment (polomani) [reply src/resources/assets/assetsApi.ts]: you can come up with better name then `AssetProps` ofc

File: src/resources/classes/asset.ts [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+import { AssetList } from './assetList';
+
+export interface SubtreeOptions {
+  depth?: number;
+}
+export interface DeleteOptions {
+  recursive?: boolean;
+}
+export class Asset implements types.Asset {
+  public id: types.CogniteInternalId;
```
Comment (polomani) [inline src/resources/classes/asset.ts]: the same about types in this file

File: src/resources/classes/asset.ts [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        {
+          id: this.id,
+        },
+      ],
+      options
+    );
+  };
+
+  public parent: () => Promise<null | Asset> = async () => {
+    const { parentId } = this;
```
Comment (polomani) [inline src/resources/classes/asset.ts]: destructive assignment for only one property?
maybe just `if (this.parentId) {...} else { return null }`

File: src/resources/classes/asset.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+      .autoPagingToArray({ limit: Infinity });
+    return new AssetList(this.client, childAssets);
+  };
+
+  public subtree = async (options?: SubtreeOptions) => {
+    const query: SubtreeOptions = options || {};
+    return this.client.assets.retrieveSubtree(this.id, query.depth || Infinity);
+  };
+
+  public timeSeries = async (filter?: types.TimeseriesFilter) => {
```
Comment (polomani) [inline src/resources/classes/asset.ts]: how about
```suggestion
  public timeSeries = async (filter: types.TimeseriesFilter = {}) => {
```
and then
```
.list({
        ...filter,
        assetIds: [this.id],
      })
```
it's just `...(filter || {})` looks cryptic for me.
but you decide

Comment (polomani) [reply src/resources/classes/asset.ts]: not a biggie

File: src/resources/classes/asset.ts [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+        ...(filter || {}),
+        assetIds: [this.id],
+      })
+      .autoPagingToArray({ limit: Infinity });
+  };
+
+  public events = async (filter?: types.EventFilter) => {
+    return this.client.events
+      .list({
+        filter: { ...(filter || {}), assetIds: [this.id] },
```
Comment (polomani) [inline src/resources/classes/asset.ts]: the same as above

File: src/resources/classes/asset.ts [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+      })
+      .autoPagingToArray({ limit: Infinity });
+  };
+
+  public files = async (filter?: types.FileFilter) => {
+    return this.client.files
+      .list({
+        filter: { ...(filter || {}), assetIds: [this.id] },
```
Comment (polomani) [inline src/resources/classes/asset.ts]: same as above

File: src/resources/classes/assetList.ts [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+
+export class AssetList extends Array<Asset> {
+  private client: CogniteClient;
+  constructor(client: CogniteClient, items: Asset[]) {
+    super(...items);
+    this.client = client;
+  }
+
+  public delete = () => {
+    return this.client.assets.delete(this.map(asset => ({ id: asset.id })));
```
Comment (polomani) [inline src/resources/classes/assetList.ts:30]: I would say that this function should have async/await as well

File: src/resources/classes/assetList.ts [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+  };
+
+  public timeSeries = async (): Promise<GetTimeSeriesMetadataDTO[]> => {
```
Comment (polomani) [inline src/resources/classes/assetList.ts]: I think you don't need to specify type here, as you do this `as GetTimeSeriesMetadataDTO[]` already

File: src/resources/classes/assetList.ts [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    return (await this.getResourcesFromAssets(
+      this.client.timeseries
+    )) as GetTimeSeriesMetadataDTO[];
+  };
+
+  public files = async (): Promise<FilesMetadata[]> => {
```
Comment (polomani) [inline src/resources/classes/assetList.ts]: same as above

File: src/resources/classes/assetList.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    return (await this.getResourcesFromAssets(
+      this.client.files
+    )) as FilesMetadata[];
+  };
+
+  public events = async (): Promise<CogniteEvent[]> => {
+    return this.getResourcesFromAssets(this.client.events);
```
Comment (polomani) [inline src/resources/classes/assetList.ts]: I believe you should have await here, also you can remove type above and do type casting here instead 
```
as CogniteEvent[];
```
as in other places above

File: src/standardMethods.ts [hunk 25]
```diff
@@ -11,24 +11,27 @@ import { MetadataMap } from './metadata';
 import { promiseAllWithData } from './resources/assets/assetUtils';
 import { CursorResponse, ItemsResponse } from './types/types';
 
-type CreateEndpoint<RequestType, ResponseType> = (
-  items: RequestType[]
-) => Promise<ResponseType[]>;
+export type Newable<T> = new (...args: any[]) => T;
```
Comment (polomani) [inline src/standardMethods.ts]: you never use it. remove?

**Review by polomani — CHANGES_REQUESTED** [review]
the most important are: `items => items`, `type.Type`, delays, `ts-ignore`

File: src/__tests__/standardMethods.spec.ts (hunk 12 above)
Comment (polomani) [reply src/__tests__/standardMethods.spec.ts]: as I suggested below

File: src/__tests__/resources/asset.integration.spec.ts (hunk 6 above)
Comment (eirikdahlen) [reply src/__tests__/resources/asset.integration.spec.ts]: Fixed with [this PR](https://github.com/cognitedata/cognitesdk-js/pull/252)

File: src/__tests__/resources/asset.integration.spec.ts (hunk 7 above)
Comment (eirikdahlen) [reply src/__tests__/resources/asset.integration.spec.ts]: For this to work I need to declare it outside of the describe for the whole test. Is that really a better way to do it?

File: src/__tests__/resources/asset.integration.spec.ts (hunk 3 above)
Comment (eirikdahlen) [reply src/__tests__/resources/asset.integration.spec.ts]: Yes, can remove

File: src/__tests__/resources/asset.integration.spec.ts (hunk 8 above)
Comment (eirikdahlen) [reply src/__tests__/resources/asset.integration.spec.ts]: Thanks!

File: src/resources/assets/assetsApi.ts (hunk 14 above)
Comment (eirikdahlen) [reply src/resources/assets/assetsApi.ts:148]: Isn't this way more explainable? Then you don't have to check `getAssetSubtree` to understand what the 0 is.

File: src/standardMethods.ts (hunk 25 above)
Comment (eirikdahlen) [reply src/standardMethods.ts]: It was something that were copied from the files Fredrik had already started on, so I don't know what it does.

File: src/resources/assets/assetsApi.ts (hunk 14 above)
Comment (polomani) [reply src/resources/assets/assetsApi.ts:148]: agree

File: src/__tests__/resources/asset.integration.spec.ts (hunk 7 above)
Comment (polomani) [reply src/__tests__/resources/asset.integration.spec.ts]: what if you move `async function testResourceType` inside `describe('Asset', () => {` and all `const asset =` on the same level?
As you don't use `testResourceType` anywhere outside of that describe, I don't see the reason to pass this param to it

File: src/standardMethods.ts (hunk 25 above)
Comment (polomani) [reply src/standardMethods.ts]: yeah, but if he needs it he will probably add it to his pr

File: src/resources/3d/revisions3DApi.ts [hunk 26]
```diff
@@ -43,7 +43,7 @@ export class Revisions3DAPI {
    * ```
    */
   public list: Revisions3DListEndpoint = (modelId, filter) => {
-    return generateListEndpoint<Revision3DListRequest, Revision3D>(
+    return generateListEndpoint<Revision3DListRequest, Revision3D, Revision3D>(
```
Comment (polomani) [inline src/resources/3d/revisions3DApi.ts]: revert this one, please

File: src/resources/3d/assetMappings3DApi.ts [hunk 27]
```diff
@@ -37,7 +37,11 @@ export class AssetMappings3DAPI {
    * ```
    */
   public list: AssetMappings3DListEndpoint = (modelId, revisionId, filter) => {
-    return generateListEndpoint<AssetMappings3DListFilter, AssetMapping3D>(
+    return generateListEndpoint<
+      AssetMappings3DListFilter,
+      AssetMapping3D,
+      AssetMapping3D
+    >(
```
Comment (polomani) [inline src/resources/3d/assetMappings3DApi.ts]: please, revert this one as well

File: src/resources/3d/viewer3DApi.ts [hunk 28]
```diff
@@ -106,7 +106,11 @@ export class Viewer3DAPI {
     revisionId,
     query
   ) => {
-    return generateListEndpoint<ListRevealSectors3DQuery, RevealSector3D>(
+    return generateListEndpoint<
+      ListRevealSectors3DQuery,
+      RevealSector3D,
+      RevealSector3D
+    >(
```
Comment (polomani) [inline src/resources/3d/viewer3DApi.ts]: please, revert this

File: src/resources/3d/revisions3DApi.ts (hunk 26 above)
Comment (eirikdahlen) [reply src/resources/3d/revisions3DApi.ts]: ✅

File: src/resources/assets/assetsApi.ts [hunk 29]
```diff
@@ -90,33 +95,121 @@ export class AssetsAPI {
...
+    const assetArray = assets.map(asset => new Asset(this.client, asset));
+    return new AssetList(this.client, assetArray);
+  };
+
+  private async getAssetSubtree(
+    assets: AssetList,
+    currentDepth: number,
+    depth: number = Infinity
+  ): Promise<AssetList> {
+    const subtree: AssetList = assets;
```
Comment (ghost) [inline src/resources/assets/assetsApi.ts]:
```suggestion
    const subtree = assets;
```

File: src/resources/classes/asset.ts [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+  AssetName,
+  AssetSource,
+  CogniteClient,
+  CogniteExternalId,
+  CogniteInternalId,
+  EventFilter,
+  FileFilter,
+  Metadata,
+  TimeseriesFilter,
+} from '../..';
```
Comment (ghost) [inline src/resources/classes/asset.ts]: Could you use explicit file path (include the filename)?

File: src/resources/classes/asset.ts [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+        filter: {
+          parentIds: [this.id],
+        },
+      })
+      .autoPagingToArray({ limit: Infinity });
+    return new AssetList(this.client, childAssets);
+  };
+
+  /**
+   * Returns the full subtree of the current asset, including the asset itself
```
Comment (ghost) [inline src/resources/classes/asset.ts:104]: You should also document the option parameter

File: src/resources/classes/asset.ts [hunk 32]
```diff
@@ -0,0 +1,157 @@
...
+  public subtree = async (options?: SubtreeOptions) => {
+    const query: SubtreeOptions = options || {};
+    return this.client.assets.retrieveSubtree(
+      { id: this.id },
+      query.depth || Infinity
+    );
+  };
+
+  /**
+   * Returns all timeseries for the current asset
```
Comment (ghost) [inline src/resources/classes/asset.ts:119]: Document the parameter

File: src/resources/classes/asset.ts [hunk 33]
```diff
@@ -0,0 +1,157 @@
...
+    return this.client.timeseries
+      .list({
+        ...filter,
+        assetIds: [this.id],
+      })
+      .autoPagingToArray({ limit: Infinity });
+  };
+
+  /**
+   * Returns all events for the current asset
```
Comment (ghost) [inline src/resources/classes/asset.ts:135]: Document the parameter

File: src/resources/classes/asset.ts [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+  public events = async (filter: EventFilter = {}) => {
+    return this.client.events
+      .list({
+        filter: { ...filter, assetIds: [this.id] },
+      })
+      .autoPagingToArray({ limit: Infinity });
+  };
+
+  /**
+   * Returns all files for the current asset
```
Comment (ghost) [inline src/resources/classes/asset.ts:150]: Document the parameter

**Review by polomani — APPROVED** [review]

**Pull request merged**